### ISSUE TITLE
Workaround for React and Google Analytics

### ISSUE THREAD

**Issue by chrisunitt-xx** [layersoflondon/application#447]
Angular doesn't play nicely with Google Analytics and Tag Manager. Without some indication that the page content has changed (usually a page load) data is not collected. At the moment GA is hugely undercounting views of content.

I've tried using the History API to pick up changes, but it's not great. In particular, I can't grab useful variables for page title and URL.

When new content is loaded dynamically (mostly when collections or records are selected on /map) we would need the following data pushed to the data layer…

```
<script>
  dataLayer = [{
    'event': 'virtualPageview',
    'virtualPageUrl': '/map/collections/19'
    'virtualPageTitle': 'J Seager and Son Bookshop 197 East India Dock Road Poplar'
  }];
</script>
```

I can then configure GTM to look for the 'virtualPageview' event and cobble together a GA pageview tag using the data from the variables.

**Comment by chrisunitt-xx** [comment]
Ah, I thought I saw an Angular file in there, my bad. It's all the same though - this is a drawback of angular, react, vue.js, and anything else that loads significant levels of content dynamically.

Sidenote: those types of js framework can do funny things to search engine crawlers too, so it may be worth checking once Search Console has updated with some data.

**Comment by chrisunitt-xx** [comment]
Soz, but I wanted to check the reason for closing this. At the moment there's no way of tracking how people are using the map and whether they're looking at records, collections, or layers. That's quite a big blindspot for the HLF evaluation.

**Comment by chrisunitt-xx** [comment]
A bit more detail. Tried to be as thorough as possible but let me know if anything's not clear or if I missed anything.

At the moment we can't track when someone:

1. views a collection or record from the /map
2. views a user's profile, or some media from a record
3. goes back to the map from a collection or record
4. uses the map search box
5. selects a layer

This is all to do with viewing/consuming content. There needs to be a similar exercise for when someone creates an account, creates a collection, uploads a record, and publishes a collection, as these things can't be tracked at the moment either. You might well get the gist from the following though.

**For 1, 2, and 3** the solution is the same. We need the page path and title pushed to the data layer so that we can use GTM to trigger a pageview tag populated with the necessary fields.

```
<script>
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({
    'event': 'virtualPageview',
    'virtualPagePath': '/map/collections/19' **// the page path**
    'virtualPageTitle': 'J Seager and Son Bookshop 197 East India Dock Road Poplar' **// the page title**
  }];
</script>
```

pageTitle is tricky for media (ie image galleries). I'd use the title of the resource and the media id bit of the URL. For instance, https://www.layersoflondon.org/map/records/5663/media/13300 would have a title of 'Christ Church, Spitalfields - media/13300'.

**For 4, the search function**

**Find a place (on /map)**. When someone clicks on a location from the suggested results…

```
<script>
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({
    'event': 'mapSearch',
    'mapSearch': '{{text of the autosuggest search result}} '
  }];
</script>
```

**Record search** from homepage or /map

```
<script>
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({
    'event': 'recordSearch',
    'recordSearchTerm': '{{search term}} '
  }];
</script>
```

**For 5, the layers**

When someone loads /map for the first time…

```
<script>
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({
    'event': 'layerSelected',
    'layerSelected': 'none' **// or whatever is active - I'm not sure if it's possible to link someone to an active layer**
  }];
</script>
```

When someone selects a new layer…

```
<script>
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({
    'event': 'layerSelected',
    'layerSelected': 'Bomb Damage (1945)' **// can be a | separated list if there are multiple selected. 
  }];
</script>
```

**Comment by edtjones** [comment]
@chrisunitt-xx does pushing onto the dataLayer work at any point during the lifecycle of a page? given it's just an array, what happens when it's really big? (lots of events recorded) Does GTM do the right thing?

**Comment by edtjones** [comment]
as an aside, I've just deployed a change so there are sensible page titles and meta description / keyword tags

**Comment by chrisunitt-xx** [comment]
Good q. For #5 the code for this should ideally go above the GTM container on the page, so that the data is available when GTM loads. For all subsequent pushes they just need to happen at the time of the interaction.

Here's an example http://recordit.co/x8wlageXvj showing the GTM preview pane on another site that has this working. Each time I click a thing there's a dataLayer push with an event key (value is virtualPageview) that shows up on the left-hand side. In response to each push, I'm firing (in this case) a pair of GA pageview tags.

And yep, the values in the dataLayer (event, virtualPagePath, virtualPageTitle, recordSearch, mapSearch, layerSelected) will be over-written every time a new value is pushed there, so it'll never be too big. Tbh, we're not really interested in persisting data in their anyway - we just fire a tag at the point some new data is pushed.

**Comment by edtjones** [comment]
Sorry for all the questions:

- can't we use an array for layerSelected? Why pipe delimited? is that easier for you?
- why do you want a layerSelected event when the map first loads?

**Comment by edtjones** [comment]
also another question - do you need exactly what's in the `<title>` attribute or are you happy with an approximation?

**Comment by chrisunitt-xx** [comment]
> can't we use an array for layerSelected? Why pipe delimited? is that easier for you?

Because the field in GA isn't smart enough to deal with an array. We can only send a string.

> why do you want a layerSelected event when the map first loads?

So for each pageview of /map we can say how the user was using it. I suppose is there's no value on load then that variable will just be 'undefined' and I can map that to 'none' in GTM. Although there's also a bit of me that's anticipating a day when you can send someone a link to an active layer, or when a selected layer survives a page refresh.

> do you need exactly what's in the <title> attribute or are you happy with an approximation?

An approximation would probably be fine. As long as it'll look sensible enough in reports, I guess.

**Comment by edtjones** [comment]
@chrisunitt-xx at your leisure can you sanity-check my work? I've deployed to production.

The map view with empty layers thing is a bit harder than I'd expected. You're right that we should have a URL which persists the state of the layers - we're doing work on refactoring layers next week so I'll roll it into that. For now, can you check for nullness?

**Comment by edtjones** [comment]
also to say - for reasons I don't understand I was getting a silly problem with `dataLayer` not being mutated, so I've renamed it `dataLayerParams` and it's working properly (and I've updated the gtm snippet to look for that array name)

**Comment by edtjones** [comment]
@chrisunitt-xx I THINK I've done everything you need on this ticket so I'm optimistically closing it ;) Let me know if something isn't right.

**Comment by chrisunitt-xx** [comment]
Hi Ed. Thanks, I just worked through all of this. It's definitely a big step forward!

There are a couple of quirks and I've added workarounds where possible. Items 1 and 3b might need a small bit of work, I've found a workaround for 2, and 3a would be ridiculously quick to solve.

Details below. Cheers!

**1. Virtual pageviews on /map**
When someone returns to /map from a collection/record/user the virtualPageview fires three times (once immediately, then twice more when collections and pins are loaded on the page). Also, the value for virtualPageTitle is retained from whichever collection/record/etc had been looked at before.

In the short-term I've told GTM to ignore virtualPageviews where the page path is /map so we don't massively over-count. It'd be good to fix this, so that the values only fired once and looked something like this…

```
<script>
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({
    'event': 'virtualPageview',
    'virtualPagePath': '/map
    'virtualPageTitle': 'Layers of London - map'
  }];
</script>
```

**2. Virtual pageview on media pages**
These are firing twice - once with the '- media/13334' appended to the title, and once without. I've added a thing to the trigger to stop it recording virtual pageviews that don't include '- media/', so I can workaround this.

**3. Search**

_a) Record search_
There's a dataLayer push working for this on the results page from the homepage or sidebar (ie https://www.layersoflondon.org/map/search?show_results=true&q=angel&&), but the event says 'mapSearch' instead of 'recordSearch'.

_b) Map search_
There's no dataLayer push for when people select a location from this.
